src/docfx/build/zonepivot/ZonePivotProvider.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Docs.Build
+{
+    internal class ZonePivotProvider
+    {
+        private readonly ErrorBuilder _errors;
+        private readonly MetadataProvider _metadataProvider;
+        private readonly Input _input;
+        private readonly Lazy<PublishUrlMap> _publishUrlMap;
+
+        private readonly ConcurrentDictionary<FilePath, FilePath?> _zonePivotDefinitionFileCache = new ConcurrentDictionary<FilePath, FilePath?>();
+        private readonly ConcurrentDictionary<FilePath, ZonePivotGroupDefinitionModel?> _zonePivotDefinitionModelCache =
+            new ConcurrentDictionary<FilePath, ZonePivotGroupDefinitionModel?>();
+
+        public ZonePivotProvider(
+            ErrorBuilder errors,
+            MetadataProvider metadataProvider,
+            Input input,
+            Lazy<PublishUrlMap> publishUrlMap)
+        {
+            _errors = errors;
+            _metadataProvider = metadataProvider;
+            _input = input;
+            _publishUrlMap = publishUrlMap;
+        }
+
+        public (FilePath?, ZonePivotGroup?) GetZonePivotGroup(FilePath file)
```

Comment:
Change the return type to ensure that either there is no pivot group (`null`), or there is a valid one (both non-null). Internally they could share a single dictionary.
```suggestion
        public (FilePath, ZonePivotGroup)? TryGetZonePivotGroup(FilePath file)
```